### PR TITLE
Migrate master to MN4

### LINKED ISSUE #282: Update to MN 4
### Issue description

We have to set `master` as MN 4 compatible.

### PR DESCRIPTION
**Pull request by wetted**
closes #282

### CODE REVIEW

**Comment by wetted** [comment]
@harissecic Can you tell why these tests are failing now? I am on a M1 mac so – unfortunately – I cannot run the tests: they try to bind to a protected directory on my mac, even if I run them using testcontainers.cloud. As a result the testcontainer fails to startup.

I have pretty everything migrated to enable this as a module for Micronaut 4, including all the renovate dependendency updates, and fixes for depreacations introduced by Micronaut 4 core.

File: build.gradle [hunk 1]
```diff
@@ -1,4 +1,10 @@
-plugins {
-    id 'io.micronaut.build.internal.docs'
-    id 'io.micronaut.build.internal.quality-reporting'
+//plugins {
```
Comment (sdelamo) [inline build.gradle:1]: if it is not necessary remove commented out code.

File: buildSrc/src/main/groovy/io.micronaut.build.internal.pulsar-base.gradle [hunk 2]
```diff
@@ -0,0 +1,11 @@
+repositories {
+    mavenCentral()
+    maven { url "https://s01.oss.sonatype.org/content/repositories/snapshots/" }
+}
+
+//configurations.all {
```
Comment (sdelamo) [inline buildSrc/src/main/groovy/io.micronaut.build.internal.pulsar-base.gradle:6]: if it is not necessary remove commented out code.

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -1,46 +1,61 @@
...
 
 conscrypt-openjdk = "2.5.2"
-kotlin = '1.7.20'
-kotest = '4.6.4'
-testcontainers = '1.17.5'
-junit = '5.9.1'
-
-micronaut = '3.8.2'
-micronaut-serde = '1.3.2'
+groovy = "4.0.7"
```
Comment (sdelamo) [inline gradle/libs.versions.toml:4]:
```suggestion
groovy = "4.0.8"
```

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -1,46 +1,61 @@
...
+
 micronaut-docs = { module = "io.micronaut.docs:micronaut-docs-asciidoc-config-props", version.ref = "micronaut-docs" }
 
-# This can be replaced with the Micronaut BOM version once the base Micronaut version is bumped to 3.5
-micronaut-grpc-protobuf-support = "io.micronaut.grpc:micronaut-protobuff-support:3.3.1"
+gradle-micronaut = { module = "io.micronaut.gradle:micronaut-gradle-plugin", version.ref = "micronaut-gradle-plugin" }
+gradle-kotlin = { module = "org.jetbrains.kotlin:kotlin-gradle-plugin", version.ref = "kotlin" }
+
+#[plugins]
+#kotlin-jvm = { id = 'org.jetbrains.kotlin.jvm', version.ref = 'kotlin' }
```
Comment (sdelamo) [inline gradle/libs.versions.toml:57]: if not necessary remove commented out code

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
@wetted it fails in CI

**Comment by harissecic** [comment]
> @harissecic Can you tell why these tests are failing now? I am on a M1 mac so – unfortunately – I cannot run the tests: they try to bind to a protected directory on my mac, even if I run them using testcontainers.cloud. As a result the testcontainer fails to startup.
> 
> I have pretty everything migrated to enable this as a module for Micronaut 4, including all the renovate dependendency updates, and fixes for depreacations introduced by Micronaut 4 core.

For some reason beans are not initialized for tests. In the same class that's failing, for example:
```
PulsarConsumerSpec > test consumer read default topic and array FAILED |  
  | io.micronaut.context.exceptions.NoSuchBeanException at PulsarConsumerSpec.groovy:48
```
there's a bean at the bottom. Not sure if upgrades made my code not see test defined inner classes in groovy or is it because of Java modules, or maybe `spec.name` should be changed. I need to investigate.

For  your case it might be that I need to change tests (not sure how) to somehow generate file in the test container for Pulsar which should have replaced some values for configuring broker and client (StandaloneConf for pulsar with all-in-one container and ClientConf for brokers communicating with each other in test-share-module). My problem is that I had issues with files in the JAR and moving those to containers so I let Java tests create temp files and copy them to Pulsar container. If you have a better way of doing this let me know. Maybe I don't need to change these values and I can somehow have hardcoded files but I still need a way to copy them to container without issues on Linux based and Windows OS. Windows is probably least important since most devs use Mac or Linux (or WSL on Windows if not directly Linux) but I don't have a Mac to test different setups for file copy.

Anyway I will investigate but I would require some help on how to copy file from build folder to container under TestContainers so that it works on all platforms.

**Comment by harissecic** [comment]
@wetted I've managed to fix Micronaut not finding beans in test which was due to removal of some dependencies. Now their back (annotation processing). I've also changed how files work so maybe you'll be able to use it on your Mac now. One thing I still need to fix but not sure how is missing `JsonMapper` from Micronaut. Bean is not available and I don't know which dependency should be added to configure it. Easiest solution would be to instantiate it locally and add to application context in tests but I don't want to do that if there's a dependency that does this from micronaut-serde or such

**Comment by wetted** [comment]
@harissecic The reason I cannot run the tests try to bind to my mac at `/var/folders/7w/q4klm5d94hxcb_5dtxqmq84w0000gp/T/5393098636774802321.tmp\` but that's a protected directory, so the container fails to start.

```
java.lang.ExceptionInInitializerError
	at java.base/java.lang.Class.forName(Class.java:467)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at io.micronaut.pulsar.PulsarAwareTest.setupSpec(PulsarAwareTest.groovy:32)
Caused by: java.lang.Exception: 
	at io.micronaut.pulsar.shared.PulsarTls.<clinit>(PulsarTls.groovy:60)
	... 3 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed
	at app//org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:349)
	at app//org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
	at app//io.micronaut.pulsar.shared.PulsarTls.<clinit>(PulsarTls.groovy:58)
	... 3 more
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at app//org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
	at app//org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:334)
	... 5 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
	at app//org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:542)
	at app//org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:344)
	at app//org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	... 6 more
Caused by: com.github.dockerjava.api.exception.BadRequestException: Status 400: {"message":"failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/var/folders/7w/q4klm5d94hxcb_5dtxqmq84w0000gp/T/5393098636774802321.tmp\" to rootfs at \"/pulsar/conf/client.conf\": mount /var/folders/7w/q4klm5d94hxcb_5dtxqmq84w0000gp/T/5393098636774802321.tmp:/pulsar/conf/client.conf (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type"}
```

**Comment by harissecic** [comment]
> @harissecic The reason I cannot run the tests try to bind to my mac at `/var/folders/7w/q4klm5d94hxcb_5dtxqmq84w0000gp/T/5393098636774802321.tmp\` but that's a protected directory, so the container fails to start. See the last line of the following stack trace.
> 
> ```
> java.lang.ExceptionInInitializerError
> 	at java.base/java.lang.Class.forName(Class.java:467)
> 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
> 	at io.micronaut.pulsar.PulsarAwareTest.setupSpec(PulsarAwareTest.groovy:32)
> Caused by: java.lang.Exception: 
> 	at io.micronaut.pulsar.shared.PulsarTls.<clinit>(PulsarTls.groovy:60)
> 	... 3 more
> Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed
> 	at app//org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:349)
> 	at app//org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
> 	at app//io.micronaut.pulsar.shared.PulsarTls.<clinit>(PulsarTls.groovy:58)
> 	... 3 more
> Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
> 	at app//org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
> 	at app//org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:334)
> 	... 5 more
> Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
> 	at app//org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:542)
> 	at app//org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:344)
> 	at app//org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
> 	... 6 more
> Caused by: com.github.dockerjava.api.exception.BadRequestException: Status 400: {"message":"failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/var/folders/7w/q4klm5d94hxcb_5dtxqmq84w0000gp/T/5393098636774802321.tmp\" to rootfs at \"/pulsar/conf/client.conf\": mount /var/folders/7w/q4klm5d94hxcb_5dtxqmq84w0000gp/T/5393098636774802321.tmp:/pulsar/conf/client.conf (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type"}
> ```

Latest push from me to this branch should avoid temp files and allow file from gradle resource build dir to just change permission to rwx for all. Not sure will it fix this but it should be a different path at least

**Comment by wetted** [comment]
> Latest push from me to this branch should avoid temp files and allow file from gradle resource build dir to just change permission to rwx for all. Not sure will it fix this but it should be a different path at least

I will check. I also think I know what's causing some test failures (I missed something in the migration). I will grab your changes and see if I can fix it.

**Comment by wetted** [comment]
@harissecic your changes don't seem to help, although the stacktrace shows the changed directories.

```
Caused by: com.github.dockerjava.api.exception.BadRequestException: Status 400: {"message":"failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/Users/wetted/Development/projects/micronaut/micronaut-pulsar/test-suite/test-pulsar-module/build/resources/test/standalone.conf\" to rootfs at \"/pulsar/conf/standalone.conf\": mount /Users/wetted/Development/projects/micronaut/micronaut-pulsar/test-suite/test-pulsar-module/build/resources/test/standalone.conf:/pulsar/conf/standalone.conf (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type"}

```

**Comment by wetted** [comment]
I still can't run tests locally, but I fixed them for CI and they all now pass. This is ready for review and merge, but running on M1 Mac with Ventura 13.2 still doesn't work for me.

**Review by harissecic — APPROVED** [review]
This is fine for now I think. It's only tests for local dev on Mac or tightened security systems that are going to fail. I've added file permission change into the test for standalone.conf and client.conf from resource build directory. On Mac I guess you have to explicitly add those files to shared part in Docker - or use Linux VM.

**Comment by wetted** [comment]
> This is fine for now I think. It's only tests for local dev on Mac or tightened security systems that are going to fail. I've added file permission change into the test for standalone.conf and client.conf from resource build directory. On Mac I guess you have to explicitly add those files to shared part in Docker - or use Linux VM.

Thank you for your contribution and feedback.

**Pull request merged**